### PR TITLE
Add "--list-languages" command line argument

### LINKED ISSUE #69: Add option to display all supported languages
Maybe:
```
> bat --list-languages
ActionScript			as
AppleScript				applescript
ASP						asa
Batch File				bat, cmd
C#						cs, csx
...
```

See: https://docs.rs/syntect/2.0.1/syntect/parsing/struct.SyntaxSet.html#method.syntaxes

### PR DESCRIPTION
**Pull request by connorkuehl**
Closes #69 

```
$ ./target/debug/bat.exe Cargo.toml --list-languages
ASP                              | asa
HTML (ASP)                       | asp
ActionScript                     | as
AppleScript                      | applescript, script editor
Batch File                       | bat, cmd
NAnt Build File                  | build
C#                               | cs, csx
C++                              | cpp, cc, cp, cxx, c++, C, h, hh, hpp, hxx, h++, inl, ipp
C                                | c, h
--snipped--
```

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -437,6 +437,12 @@ fn run() -> Result<()> {
                 .default_value("auto")
                 .help("When to use the pager"),
         )
+        .arg(
+            Arg::with_name("list languages")
+                .long("list-languages")
+                .takes_value(false)
```
Comment (sharkdp) [inline src/main.rs]: I think `false` is the default value, so `.takes_value` can be removed here.

File: src/main.rs [hunk 2]
```diff
@@ -437,6 +437,12 @@ fn run() -> Result<()> {
                 .default_value("auto")
                 .help("When to use the pager"),
         )
+        .arg(
+            Arg::with_name("list languages")
```
Comment (sharkdp) [inline src/main.rs]: I'd prefer if we use list-languages here, just like the `.long(..)` name.

File: src/main.rs [hunk 3]
```diff
@@ -486,6 +492,31 @@ fn run() -> Result<()> {
                 )
             })?;
 
+            if let Some(_) = app_matches.values_of("list languages") {
```
Comment (sharkdp) [inline src/main.rs]: You can use `if app_matches.is_present("list-languages")`.

File: src/main.rs [hunk 4]
```diff
@@ -486,6 +492,31 @@ fn run() -> Result<()> {
...
+                let languages = assets.syntax_set.syntaxes();
+
+                let longest = match languages.iter()
+                    .map(|s| s.name.len())
+                    .max() {
```
Comment (sharkdp) [inline src/main.rs]: you can use `languages.iter().....max().unwrap_or(32)` here instead of the `match` statement.

File: src/main.rs [hunk 5]
```diff
@@ -486,6 +492,31 @@ fn run() -> Result<()> {
...
+                    Some(length) => length,
+                    None => 32, // Fallback width if they have no language definitions.
+                };
+
+                for lang in languages {
+                    print!("{:width$} | ", lang.name, width = longest);
```
Comment (sharkdp) [inline src/main.rs]: Nice! Didn't know this was possible.

File: src/main.rs [hunk 6]
```diff
@@ -486,6 +492,31 @@ fn run() -> Result<()> {
...
+                    for i in 0..lang.file_extensions.len() {
```
[the remark this answers is not in the document]
Comment (sharkdp) [reply src/main.rs]: I think you should be able to use `lang.file_extensions.join(", ")`.

[the remark this answers is not in the document]
Comment (connorkuehl) [reply src/main.rs]: That worked. Thank you!

**Comment by connorkuehl** [comment]
Yes, we should limit the size of the extensions column. I had no idea some languages could have so many extensions lol.

I implemented simple line-wrapping for it. I'm still trying to learn more Rust, so if it is unnecessarily complex I would love some feedback to trim it down. I tried brainstorming a cool and simple iterator solution to this problem but didn't get too far.

Here's what it looks like:

```
--snipped--
reStructuredText                 | rst, rest
Ruby                             | rb, Appfile, Appraisals, Berksfile, Brewfile, capfile, cgi,
                                 | Cheffile, config.ru, Deliverfile, Fastfile, fcgi, Gemfile,
                                 | gemspec, Guardfile, irbrc, jbuilder, podspec, prawn, rabl,
                                 | rake, Rakefile, Rantfile, rbx, rjs, ruby.rail, Scanfile,
                                 | simplecov, Snapfile, thor, Thorfile, Vagrantfile
Cargo Build Results              |
Rust                             | rs
SQL                              | sql, ddl, dml
--snipped--
```

File: src/main.rs [hunk 7]
```diff
@@ -486,6 +491,43 @@ fn run() -> Result<()> {
...
+                    .map(|s| s.name.len())
+                    .max()
+                    .unwrap_or(32); // Fallback width if they have no language definitions.
+
+                let separator = " | ";
+                for lang in languages {
+                    print!("{:width$}{}", lang.name, separator, width = longest);
+
+                    // Line-wrapping for the possible file extension overflow.
+                    let desired_width = 48;
```
Comment (sharkdp) [inline src/main.rs]: Maybe this could be `desired_width = term_width - longest - padding`?

Comment (connorkuehl) [reply src/main.rs]: How should I get the term_width? I looked around and I'm not seeing a function I can call to get that value.

I notice it's set with a literal `90` in `run()`; should I just create a constant and apply it to both places?

Comment (sharkdp) [reply src/main.rs]: The `term_width` is computed/retrieved in `printer.s`. We should probably move that to the `run` function and store it in the `Options` struct.

**Comment by sharkdp** [comment]
We should probably also filter out Languages without any extensions. These are typically syntax definitions which are included by others, e.g. "Regular Expressions (Python)"

**Comment by sharkdp** [comment]
Actually, we should probably just hide the ones that have `hidden == true` (https://docs.rs/syntect/2.0.1/syntect/parsing/syntax_definition/struct.SyntaxDefinition.html)

File: src/main.rs [hunk 8]
```diff
@@ -559,6 +566,46 @@ fn run() -> Result<()> {
...
+
+                let separator = " | ";
+                for lang in languages {
+                    if lang.hidden {
+                        continue;
+                    }
+                    print!("{:width$}{}", lang.name, separator, width = longest);
+
+                    // Line-wrapping for the possible file extension overflow.
+                    let desired_width = options.term_width - longest;
```
Comment (sharkdp) [inline src/main.rs]: Don't we have to subtract `separator.length()` as well?

File: src/printer.rs [hunk 9]
```diff
@@ -18,8 +17,7 @@ pub struct Printer<'a> {
 
 impl<'a> Printer<'a> {
     pub fn new(handle: &'a mut Write, options: &'a Options) -> Self {
-        let (_, term_width) = Term::stdout().size();
-        let term_width = term_width as usize;
+        let term_width = options.term_width;
```
Comment (BrainMaestro) [inline src/printer.rs]: Just replace all the usages of `term_width` with `options.term_width` then

File: src/main.rs [hunk 10]
```diff
@@ -559,6 +566,46 @@ fn run() -> Result<()> {
...
+                    // Number of characters on this line so far, wrap before `desired_width`
+                    let mut num_chars = 0;
+
+                    let mut extension = lang.file_extensions.iter().peekable();
+                    while let Some(word) = extension.next() {
+                        // If we can't fit this word in, then create a line break and align it in.
+                        if word.len() + num_chars >= desired_width {
```
Comment (sharkdp) [inline src/main.rs]: Do we have to check for `word.len() + 2 + num_chars >= ..` here due to the ", " part which is possibly appended?

File: src/printer.rs [hunk 11]
```diff
@@ -30,7 +26,7 @@ impl<'a> Printer<'a> {
         Printer {
             handle,
             colors,
-            term_width,
+            term_width: options.term_width,
```
Comment (sharkdp) [inline src/printer.rs]: I think `term_width` can be removed from `Printer` if it is already included in `options`.

**Comment by connorkuehl** [comment]
Thanks for putting up with me! I appreciate it, I learned a lot.

File: src/main.rs [hunk 12]
```diff
@@ -559,6 +566,47 @@ fn run() -> Result<()> {
...
+                    let comma_separator = ", ";
+                    let mut extension = lang.file_extensions.iter().peekable();
+                    while let Some(word) = extension.next() {
+                        // If we can't fit this word in, then create a line break and align it in.
+                        if word.len() + num_chars + comma_separator.len() >= desired_width {
+                            num_chars = 0;
+                            print!("\n{:width$}{}", "", separator, width = longest);
+                        }
+
+                        num_chars += word.len();
```
Comment (sharkdp) [inline src/main.rs]: Actually, this also needs to be `num_chars += word.len() + comma_separator.len();`. But I can also change this, if you want.

**Pull request merged**